ir/blob.go

```diff
@@ -16,6 +17,40 @@ func (b Blob) WritePretty(w io.Writer) error {
 	return err
 }
 
+func (b Blob) MarshalJSON() (bdata []byte, e error) {
+	// Temporal type to avoid recursion
+	type tmp Blob
```

Comment:
This is fine. I just don't see why you couldn't use type []byte instead of introducing a new type tmp.
(Same for other places you do this.)

Reply:
I tried, you end up inside an infinite loop because when trying to Marshal a few lines below Go enters this same function. I can share a few funny Go playgrounds with this behavior :)

Reply:
What I mean is:
func (b Blob) MarshalJSON() (bdata []byte, e error) {
  c := struct {
		Type  MarshalType `json:"type"`
		Value []byte         `json:"value"`
	}{Type: BlobType, Value: b.Bytes}
	return json.Marshal(&c)
}

Reply:
Oh, I see! Doing so breaks the unmarshalling of Pairs, because json.Unmarshal doesn't know what type to use. The fact that we use String{Value: string}, Blob{Value: []byte{}} makes this easy for these types. And this is why I need to use the "ugly" if-else in Float and Int, to accommodate for the fact that we don't use the internal "Value" in number types.

However, I suspect all of this is linked to the top-level interface thing you mention in your last comment.